presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -15,11 +15,153 @@
+            }
+        }
+
+        return new PageProcessorOutput(
+                page.getRetainedSizeInBytes(),
+                new PositionsPageProcessorIterator(session, page, SelectedPositions.positionsRange(0, page.getPositionCount())));
+    }
+
+    private class PositionsPageProcessorIterator
+            extends AbstractIterator<Page>
+    {
+        private final ConnectorSession session;
+        private final Page page;
+
+        private SelectedPositions selectedPositions;
+        private final Block[] previouslyComputedResults;
+
+        public PositionsPageProcessorIterator(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+        {
+            this.session = session;
+            this.page = page;
+            this.selectedPositions = selectedPositions;
+            this.previouslyComputedResults = new Block[projections.size()];
+        }
+
+        @Override
+        protected Page computeNext()
+        {
+            while (true) {
+                if (selectedPositions.isEmpty()) {
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                try {
+                    Page page = processBatch(batchSize);
+
+                    // remove batch from selectedPositions and previouslyComputedResults
+                    selectedPositions = selectedPositions.subRange(batchSize, selectedPositions.size());
+                    for (int i = 0; i < previouslyComputedResults.length; i++) {
+                        if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                            previouslyComputedResults[i] = previouslyComputedResults[i].getRegion(batchSize, previouslyComputedResults[i].getPositionCount() - batchSize);
+                        }
+                        else {
+                            previouslyComputedResults[i] = null;
+                        }
+                    }
+
+                    return page;
+                }
+                catch (OptimisticProjectException e) {
+                    // the page buffer filled up, so halve the batch size and retry
+                    verify(batchSize > 1);
+                    projectBatchSize = projectBatchSize / 2;
+                }
+            }
+        }
+
+        private Page processBatch(int batchSize)
+                throws OptimisticProjectException
+        {
+            Block[] blocks = new Block[projections.size()];
+
+            int pageSize = 0;
+            SelectedPositions positionsBatch = selectedPositions.subRange(0, batchSize);
+            for (int i = 0; i < projections.size(); i++) {
+                if (positionsBatch.size() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
+                    throw new OptimisticProjectException();
```

Comment:
Is it really required to control program flow with exception? Why not just return some kind of status with page attached?